Electra epoch processing

michaelsproul opened the pull request:
## Proposed Changes

Integrate Electra epoch processing with single-pass. This required a few tricks. I hope they're correct :crossed_fingers:

eserilev committed:
Attestation superstruct changes for EIP 7549 (#5644)

* update

* experiment

* superstruct changes

* revert

* superstruct changes

* fix tests

* indexed attestation

* indexed attestation superstruct

* updated TODOs

ethDreamer committed:
`superstruct` the `AttesterSlashing` (#5636)

* `superstruct` Attester Fork Variants

* Push a little further

* Deal with Encode / Decode of AttesterSlashing

* not so sure about this..

* Stop Encode/Decode Bounds from Propagating Out

* Tons of Changes..

* More Conversions to AttestationRef

* Add AsReference trait (#15)

* Add AsReference trait

* Fix some snafus

* Got it Compiling! :D

* Got Tests Building

* Get beacon chain tests compiling

---------

Co-authored-by: Michael Sproul <micsproul@gmail.com>

ethDreamer committed:
Merge remote-tracking branch 'upstream/unstable' into electra_attestation_changes

ethDreamer committed:
Make EF Tests Fork-Agnostic (#5713)

ethDreamer committed:
Finish EF Test Fork Agnostic (#5714)

ethDreamer committed:
Superstruct `AggregateAndProof` (#5715)

* Upgrade `superstruct` to `0.8.0`

* superstruct `AggregateAndProof`

realbigsean committed:
Merge remote-tracking branch 'sigp/unstable' into electra_attestation_changes

realbigsean committed:
cargo fmt

realbigsean committed:
Merge pull request #5726 from realbigsean/electra_attestation_changes

Merge unstable into Electra attestation changes

realbigsean committed:
process withdrawals updates

realbigsean committed:
cleanup withdrawals processing

realbigsean committed:
update `process_operations` deposit length check

michaelsproul committed:
Electra epoch processing

realbigsean committed:
add apply_deposit changes

realbigsean committed:
add execution layer withdrawal request processing

realbigsean committed:
process deposit receipts

realbigsean committed:
add consolidation processing

realbigsean committed:
update process operations function

realbigsean committed:
exit updates

realbigsean committed:
clean up

realbigsean committed:
update slash_validator

eserilev committed:
EIP7549 `get_attestation_indices` (#5657)

* get attesting indices electra impl

* fmt

* get tests to pass

* fmt

* fix some beacon chain tests

* fmt

* fix slasher test

* fmt got me again

* fix more tests

* fix tests

ethDreamer committed:
Some small changes (#5739)

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

ethDreamer committed:
cargo fmt (#5740)

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

realbigsean committed:
fix attestation verification

michaelsproul committed:
Sketch op pool changes

realbigsean committed:
fix get attesting indices (#5742)

* fix get attesting indices

* better errors

* fix compile

* only get committee index once

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

realbigsean committed:
Ef test fixes (#5753)

* attestation related ef test fixes

* delete commented out stuff

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

ethDreamer committed:
Fix Aggregation Pool for Electra (#5754)

* Fix Aggregation Pool for Electra

* Remove Outdated Interface

realbigsean committed:
fix ssz (#5755)

ethDreamer committed:
Get `electra_op_pool` up to date (#5756)

* fix get attesting indices (#5742)

* fix get attesting indices

* better errors

* fix compile

* only get committee index once

* Ef test fixes (#5753)

* attestation related ef test fixes

* delete commented out stuff

* Fix Aggregation Pool for Electra (#5754)

* Fix Aggregation Pool for Electra

* Remove Outdated Interface

* fix ssz (#5755)

---------

Co-authored-by: realbigsean <sean@sigmaprime.io>

ethDreamer committed:
Revert "Get `electra_op_pool` up to date (#5756)" (#5757)

This reverts commit ab9e58aa3d0e6fe2175a4996a5de710e81152896.

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into electra_op_pool

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

eserilev committed:
Compute on chain aggregate impl (#5752)

* add compute_on_chain_agg impl to op pool changes

* fmt

* get op pool tests to pass

realbigsean committed:
update the naive agg pool interface (#5760)

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

michaelsproul committed:
Fix bugs in cross-committee aggregation

michaelsproul committed:
Add comment to max cover optimisation

michaelsproul committed:
Fix assert

dapplion reviewed (approved):
Approach is solid and should have minimal overhead in practice. We can re-order maybe some things latter for better clarity but can figure it out after the interop

realbigsean committed:
Merge pull request #5749 from sigp/electra_op_pool

Optimise Electra op pool aggregation

realbigsean committed:
don't fail on empty consolidations

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

realbigsean committed:
Merge branch 'block-processing-electra' of https://github.com/sigp/lighthouse into electra-epoch-proc

realbigsean committed:
update committee offset

realbigsean committed:
update committee offset

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

realbigsean committed:
only increment the state deposit index on old deposit flow

realbigsean committed:
Merge branch 'block-processing-electra' of https://github.com/sigp/lighthouse into electra-epoch-proc

realbigsean committed:
use correct max eb in epoch cache initialization

realbigsean committed:
drop initiate validator ordering optimization

realbigsean committed:
fix initiate exit for single pass

ethDreamer committed:
Fix Electra Fork Choice Tests (#5764)

ethDreamer committed:
Fix Electra Fork Choice Tests (#5764)

ethDreamer committed:
Fix Electra Fork Choice Tests (#5764)

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

realbigsean committed:
Merge branch 'block-processing-electra' of https://github.com/sigp/lighthouse into electra-epoch-proc

realbigsean committed:
add consolidations to merkle calc for inclusion proof

ethDreamer committed:
Fix Consolidation Sigs & Withdrawals

realbigsean committed:
Merge pull request #5766 from ethDreamer/two_fixes

Fix Consolidation Sigs & Withdrawals

realbigsean committed:
Merge branches 'block-processing-electra' and 'electra-epoch-proc' of https://github.com/sigp/lighthouse into electra-epoch-proc

pawanjay176 committed:
Send unagg attestation based on fork

eserilev committed:
Subscribe to the correct subnets for electra attestations (#5782)

* subscribe to the correct att subnets for electra

* subscribe to the correct att subnets for electra

eserilev committed:
Subscribe to the correct subnets for electra attestations (#5782)

* subscribe to the correct att subnets for electra

* subscribe to the correct att subnets for electra

eserilev committed:
Subscribe to the correct subnets for electra attestations (#5782)

* subscribe to the correct att subnets for electra

* subscribe to the correct att subnets for electra

realbigsean committed:
cargo fmt

realbigsean committed:
cargo fmt

realbigsean committed:
cargo fmt

ethDreamer committed:
Publish all aggregates

ethDreamer committed:
just one more check bro plz..

ethDreamer committed:
Fix Bug In Block Processing with 0x02 Credentials

realbigsean committed:
fix slashing handling

ethDreamer committed:
Merge remote-tracking branch 'upstream/unstable'

realbigsean committed:
Merge pull request #5832 from ethDreamer/electra_attestation_changes_merge_unstable

Merge `unstable` into `electra_attestation_changes`

realbigsean committed:
Merge pull request #5835 from realbigsean/fix-validator-logic

Fix validator logic

realbigsean committed:
Merge pull request #5816 from realbigsean/electra-attestation-slashing-handling

Electra slashing handling

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

realbigsean committed:
Merge branch 'block-processing-electra' of https://github.com/sigp/lighthouse into electra-epoch-proc

ethDreamer committed:
Electra attestation changes rm decode impl (#5856)

* Remove Crappy Decode impl for Attestation

* Remove Inefficient Attestation Decode impl

* Implement Schema Upgrade / Downgrade

* Update beacon_node/beacon_chain/src/schema_change/migration_schema_v20.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

---------

Co-authored-by: Michael Sproul <micsproul@gmail.com>

eserilev committed:
Fix failing attestation tests and misc electra attestation cleanup (#5810)

* - get attestation related beacon chain tests to pass
- observed attestations are now keyed off of data + committee index
- rename op pool attestationref to compactattestationref
- remove unwraps in agg pool and use options instead
- cherry pick some changes from ef-tests-electra

* cargo fmt

* fix failing test

* Revert dockerfile changes

* make committee_index return option

* function args shouldnt be a ref to attestation ref

* fmt

* fix dup imports

---------

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

realbigsean committed:
fix some todos (#5817)

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into electra_attestation_changes

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

realbigsean committed:
Merge branch 'block-processing-electra' of https://github.com/sigp/lighthouse into electra-epoch-proc

ethDreamer committed:
Remove Duplicate KZG Commitment Merkle Proof Code (#5874)

* Remove Duplicate KZG Commitment Merkle Proof Code

* s/tree_lists/fields/

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into electra_attestation_changes

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

realbigsean committed:
Merge branch 'block-processing-electra' of https://github.com/sigp/lighthouse into electra-epoch-proc

realbigsean committed:
fix compile

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

realbigsean committed:
Merge branch 'block-processing-electra' of https://github.com/sigp/lighthouse into electra-epoch-proc

pawanjay176 committed:
Fix slasher tests (#5906)

* Fix electra tests

* Add electra attestations to double vote tests

michaelsproul committed:
Update superstruct to 0.8

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into electra_attestation_changes

michaelsproul committed:
Small cleanup in slasher tests

dapplion commented on consensus/state_processing/src/common/initiate_validator_exit.rs:
```diff
@@ -8,12 +8,12 @@ pub fn initiate_validator_exit<E: EthSpec>(
+    // Return if the validator already initiated exit
+    if validator.exit_epoch != spec.far_future_epoch {
+        return Ok(());
+    }
```
Why is it necessary to delay computing the exit cache now?

michaelsproul replied:
I don't know. @realbigsean deleted it in this commit: f4907ef971973b67fb692cb42a5119b5402a4371

realbigsean replied:
I needed to do this to fix some EF tests.  In Electra we are mutating the state when calculating `exit_queue_epoch` because we call `compute_exit_epoch_and_update_churn`. So we need to actually do the check for whether the validator exited before calculating `exit_queue_epoch`. 

I figured it wasn't worth trying to maintain the pre-fork optimization at the expense of clarity. Is it worth adding back or trying to find a similar post-fork optimization?

dapplion commented on consensus/state_processing/src/per_epoch_processing/single_pass.rs:
```diff
@@ -278,15 +325,56 @@ pub fn process_epoch_single_pass<E: EthSpec>(
+        let new_pending_balance_deposits = List::try_from_iter(
+            state
+                .pending_balance_deposits()?
+                .iter_from(ctxt.next_deposit_index)?
```
Note, this can use new milhouse splice functionality

dapplion commented on consensus/state_processing/src/epoch_cache.rs:
```diff
@@ -9,6 +9,7 @@ use types::{ActivationQueue, BeaconState, ChainSpec, EthSpec, ForkName, Hash256}
 pub struct PreEpochCache {
     epoch_key: EpochCacheKey,
     effective_balances: Vec<u64>,
+    total_active_balance: u64,
```
Why do we need to track total_active_balance here?

dapplion replied:
EDIT: Because now we call `process_single_effective_balance_update` twice, once on the original loop while populating `effective_balances` and a second time for consolidations. I'm not sold it's entirely necessary, but the current implementation is straightforward to follow

dapplion committed:
De-dup attestation constructor logic

dapplion committed:
Remove unwraps in Attestation construction

michaelsproul committed:
Clean up Electra observed aggregates (#5929)

* Use consistent key in observed_attestations

* Remove unwraps from observed aggregates

dapplion committed:
Dedup match_attestation_data

dapplion committed:
Remove outdated TODO

dapplion committed:
Use ForkName Ord in fork-choice tests

dapplion committed:
Make to_electra not fallible

dapplion committed:
Remove TestRandom impl for IndexedAttestation

dapplion committed:
Remove IndexedAttestation faulty Decode impl

dapplion committed:
Use ForkName Ord in BeaconBlockBody

dapplion committed:
Drop TestRandom impl

dapplion committed:
Add PendingAttestationInElectra

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into electra_attestation_changes

dapplion committed:
Dedup attestation constructor logic in attester cache

realbigsean committed:
Indexed att on disk (#35)

* indexed att on disk

* fix lints

* Update slasher/src/migrate.rs

Co-authored-by: ethDreamer <37123614+ethDreamer@users.noreply.github.com>

---------

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>
Co-authored-by: ethDreamer <37123614+ethDreamer@users.noreply.github.com>

dapplion committed:
Use if let Ok for committee_bits

dapplion committed:
Dedup Attestation constructor code

dapplion committed:
Diff reduction in tests

dapplion committed:
Fix beacon_chain tests

dapplion committed:
Diff reduction

dapplion committed:
Use Ord for ForkName in pubsub

dapplion committed:
Resolve into_attestation_and_indices todo

dapplion committed:
Remove stale TODO

eserilev committed:
add electra fork enabled fn to ForkName impl (#36)

* add electra fork enabled fn to ForkName impl

* remove inadvertent file

dapplion committed:
Update common/eth2/src/types.rs

Co-authored-by: ethDreamer <37123614+ethDreamer@users.noreply.github.com>

dapplion committed:
Fix beacon_chain tests

dapplion committed:
Test spec invariant

dapplion committed:
Use electra_enabled in pubsub

dapplion committed:
Remove get_indexed_attestation_from_signed_aggregate

dapplion committed:
Use ok_or instead of if let else

dapplion committed:
committees are sorted

realbigsean committed:
remove dup method `get_indexed_attestation_from_committees`

realbigsean committed:
Merge pull request #5940 from dapplion/electra_attestation_changes_lionreview

Electra attestations #5712 review

realbigsean committed:
update default persisted op pool deserialization

realbigsean committed:
ensure aggregate and proof uses serde untagged on ref

dapplion committed:
Fork aware ssz static attestation tests

eserilev committed:
Electra attestation changes from Lions review (#5971)

* dedup/cleanup and remove unneeded hashset use

* remove irrelevant TODOs

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into electra_attestation_changes

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/sigp/lighthouse into block-processing-electra

realbigsean committed:
Merge branch 'block-processing-electra' of https://github.com/sigp/lighthouse into electra-epoch-proc

michaelsproul committed:
Avoid changing slasher schema for Electra

michaelsproul committed:
Delete slasher schema v4

realbigsean committed:
Electra attestation changes sean review (#5972)

* instantiate empty bitlist in unreachable code

* clean up error conversion

* fork enabled bool cleanup

* remove a couple todos

* return bools instead of options in `aggregate` and use the result

* delete commented out code

* use map macros in simple transformations

* remove signers_disjoint_from

* get ef tests compiling

* get ef tests compiling

* update intentionally excluded files

michaelsproul committed:
Fix clippy

michaelsproul committed:
Fix compilation of beacon_chain tests

dapplion committed:
Update database.rs

dapplion committed:
Update per_block_processing.rs

dapplion committed:
Add electra lightclient types

realbigsean committed:
Update slasher/src/database.rs

realbigsean committed:
fix imports

realbigsean committed:
Merge pull request #5980 from dapplion/electra-lightclient

Add electra lightclient types

realbigsean committed:
Merge pull request #5975 from michaelsproul/electra-slasher-no-migration

Avoid changing slasher schema for Electra

realbigsean committed:
Update beacon_node/beacon_chain/src/attestation_verification.rs

realbigsean committed:
Update beacon_node/beacon_chain/src/attestation_verification.rs

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into electra_attestation_changes

realbigsean committed:
Merge branch 'electra_attestation_changes' of https://github.com/realbigsean/lighthouse into block-processing-electra

realbigsean committed:
Merge branch 'block-processing-electra' of https://github.com/sigp/lighthouse into electra-epoch-proc

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into electra-epoch-proc

realbigsean commented on consensus/state_processing/src/per_epoch_processing/single_pass.rs:
```diff
@@ -139,6 +160,16 @@ pub fn process_epoch_single_pass<E: EthSpec>(
     let mut next_epoch_cache = PreEpochCache::new_for_next_epoch(state)?;
 
+    let pending_balance_deposits_ctxt =
+        if fork_name >= ForkName::Electra && conf.pending_balance_deposits {
```
```suggestion
        if fork_name.electra_enabled() && conf.pending_balance_deposits {
```

realbigsean commented on consensus/state_processing/src/per_epoch_processing/single_pass.rs:
```diff
@@ -165,23 +196,26 @@ pub fn process_epoch_single_pass<E: EthSpec>(
-            activation_churn_limit as usize,
-        );
+
+    let mut activation_queues = if fork_name < ForkName::Electra {
```
```suggestion
    let mut activation_queues = if !fork_name.electra_enabled() {
```

realbigsean commented on consensus/state_processing/src/per_epoch_processing/single_pass.rs:
```diff
@@ -278,15 +325,56 @@ pub fn process_epoch_single_pass<E: EthSpec>(
+
+    // Process consolidations outside the single-pass loop, as they depend on balances for multiple
+    // validators and cannot be computed accurately inside the loop.
+    if fork_name >= ForkName::Electra && conf.pending_consolidations {
```
```suggestion
    if fork_name.electra_enabled() && conf.pending_consolidations {
```

realbigsean commented on consensus/state_processing/src/per_epoch_processing/single_pass.rs:
```diff
@@ -491,27 +614,77 @@ fn process_single_registry_update(
-    if exit_queue_churn >= state_ctxt.churn_limit {
-        exit_queue_epoch.safe_add_assign(1)?;
-    }
+    let exit_queue_epoch = if state_ctxt.fork_name >= ForkName::Electra {
```
```suggestion
    let exit_queue_epoch = if state_ctxt.fork_name.electra_enabled() {
```

realbigsean commented on consensus/state_processing/src/per_epoch_processing/single_pass.rs:
```diff
@@ -455,7 +543,42 @@ impl RewardsAndPenaltiesContext {
+    exit_balance_to_consume: Option<&mut u64>,
+    spec: &ChainSpec,
+) -> Result<(), Error> {
+    if state_ctxt.fork_name < ForkName::Electra {
```
```suggestion
    if !state_ctxt.fork_name.electra_enabled() {
```

realbigsean commented:
Just had some of my new favorite PR comments -> `ForkName::electra_enabled`

This looks good to me an outstanding question is whether we want to optimize this again? https://github.com/sigp/lighthouse/pull/5761#discussion_r1655028450

realbigsean committed:
Update consensus/state_processing/src/per_epoch_processing/single_pass.rs

realbigsean committed:
Update consensus/state_processing/src/per_epoch_processing/single_pass.rs

realbigsean committed:
Update consensus/state_processing/src/per_epoch_processing/single_pass.rs

realbigsean committed:
Update consensus/state_processing/src/per_epoch_processing/single_pass.rs

realbigsean committed:
Update consensus/state_processing/src/per_epoch_processing/single_pass.rs

realbigsean commented:
> This looks good to me an outstanding question is whether we want to optimize this again? [#5761 (comment)](https://github.com/sigp/lighthouse/pull/5761#discussion_r1655028450)

We can come back and optimize this if we think it's necessary

realbigsean commented:
@mergify queue
